File: Data-Structure/Python/1_Multiple_Pointers/binary-search.py
# ordered
# two pointers, start and end, move to mid

def binary_search(nums, num):
    if num < nums[0] or num > nums[-1]:
        return False
    start = 0
    end = len(nums) - 1
    while start < end:
        mid = (nums[start] + nums[end]) >> 1
        if nums[mid] == num:
            return True
        if nums[mid] > num:
            end = mid - 1
        else:
            start = mid + 1

    return False


print(binary_search([1, 2, 3, 4, 5], 6))  # False
print(binary_search([2, 3, 4, 5], 1))  # True
print(binary_search([1, 2, 3, 4, 5], 3))  # False
print(binary_search([1, 2, 4, 5], 3))  # False

File: Data-Structure/Nodejs/剑指offer专题/binary_search.js
{
    const binary_search = (start, end, key) => {
        let res = -1;  // 未搜索到数据返回-1下标
        let mid;
        while (start <= end) {
            mid = start + ((end - start) >> 1);  // 直接平均可能会溢出，所以用这个算法
            if (arr[mid] < key)
                start = mid + 1;
            else if (arr[mid] > key)
                end = mid - 1;
            else {  // 最后检测相等是因为多数搜索情况不是大于就是小于
                res = mid;
                break;
            }
        }
        return res;  // 单一出口
    }
}